frame/merkle-mountain-range/src/mmr/mmr.rs

```diff
@@ -0,0 +1,174 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+	Trait, HashingOf,
+	mmr::{
+		Node, NodeOf, Hasher,
+		storage::{Storage, OffchainStorage, RuntimeStorage},
+		utils::NodesUtils,
+	},
+	primitives,
+};
+use frame_support::{debug, RuntimeDebug};
+use sp_std::fmt;
+
+/// A wrapper around a MMR library to expose limited functionality.
+///
+/// Available functions depend on the storage kind ([Runtime](crate::mmr::storage::RuntimeStorage)
+/// vs [Off-chain](crate::mmr::storage::OffchainStorage)).
+pub struct MMR<StorageType, T, L> where
+	T: Trait,
+	L: primitives::FullLeaf,
+	Storage<StorageType, T, L>: mmr_lib::MMRStore<NodeOf<T, L>>,
+{
+	mmr: mmr_lib::MMR<
+		NodeOf<T, L>,
+		Hasher<HashingOf<T>, L>,
+		Storage<StorageType, T, L>
+	>,
+	leaves: u64,
+}
+
+impl<StorageType, T, L> MMR<StorageType, T, L> where
+	T: Trait,
+	L: primitives::FullLeaf,
+	Storage<StorageType, T, L>: mmr_lib::MMRStore<NodeOf<T, L>>,
+{
+	/// Create a pointer to an existing MMR with given number of leaves.
+	pub fn new(leaves: u64) -> Self {
+		let size = NodesUtils::new(leaves).size();
+		Self {
+			mmr: mmr_lib::MMR::new(size, Default::default()),
+			leaves,
+		}
+	}
+
+	/// Verify proof of a single leaf.
+	pub fn verify_leaf_proof(
+		&self,
+		leaf: L,
+		proof: primitives::Proof<<T as Trait>::Hash>,
+	) -> Result<bool, Error> {
+		let p = mmr_lib::MerkleProof::<
+			NodeOf<T, L>,
+			Hasher<HashingOf<T>, L>,
+		>::new(
+			self.mmr.mmr_size(),
+			proof.items.into_iter().map(Node::Hash).collect(),
+		);
+		let position = mmr_lib::leaf_index_to_pos(proof.leaf_index);
+		let root = self.mmr.get_root().map_err(|e| Error::GetRoot.debug(e))?;
+		p.verify(
```

Comment:
I guess so, yes. This will complicate the API a bit, but will make proofs tiny bit smaller. As I said in some other comment, the priority is to verify against `root`. I've added this API for completeness, but the main point of the PR is to generate the MMR root hash and generate proofs against it. Actual proof verification will happen in some other place that won't really have peaks at all.